### PR TITLE
Solves key error in creating polycyclic group, Improves _sift, induced_pcgs

### LINKED ISSUE #18804: KeyError in creating a polycylic group from Dihedral Group
I am getting this error when `n>2`and `n` is `even` is there any reason we are getting this?if yes should we raise a `exception` instead?
```
>>> G = DihedralGroup(10)
>>> A = G.polycyclic_group()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/mohit/Documents/sympy-master/sympy/sympy/combinatorics/perm_groups.py", line 4862, in polycyclic_group
    return PolycyclicGroup(pc_sequence, pc_series, relative_order, collector=None)
  File "/home/mohit/Documents/sympy-master/sympy/sympy/combinatorics/pc_groups.py", line 33, in __init__
    self.collector = Collector(self.pcgs, pc_series, relative_order) if not collector else collector
  File "/home/mohit/Documents/sympy-master/sympy/sympy/combinatorics/pc_groups.py", line 80, in __init__
    self.pc_presentation = self.pc_relators()
  File "/home/mohit/Documents/sympy-master/sympy/sympy/combinatorics/pc_groups.py", line 477, in pc_relators
    word = self.collected_word(word)
  File "/home/mohit/Documents/sympy-master/sympy/sympy/combinatorics/pc_groups.py", line 318, in collected_word
    w = self.minimal_uncollected_subword(word)
  File "/home/mohit/Documents/sympy-master/sympy/sympy/combinatorics/pc_groups.py", line 123, in minimal_uncollected_subword
    if re[index[s1]] and (e1 < 0 or e1 > re[index[s1]]-1):
KeyError: x2**4
>>> G.is_solvable
True
```

### LINKED ISSUE #18989: creating an induced polycyclic sequence from Alt(4) group takes infinite time.
```
>>> a = AlternatingGroup(4)
>>> g = a.polycyclic_group()
>>> collector = g.collector
>>> gens = [a[0],a[1]]
>>> ipcgs = collector.induced_pcgs(gens)
```
It does not evaluate.

### PR DESCRIPTION
**Pull request by mohitacecode**
#### References to other Issues or PRs

Fixes #18804 
Fixes #18989 
#### Brief description of what is fixed or changed

- Improved `collected_word`, `_sift`, `induced_pcgs`.

#### Release Notes

NO ENTRY

### CODE REVIEW

**Comment by mohitacecode** [comment]
@jksuom Please review this.

File: sympy/combinatorics/tests/test_pc_groups.py [hunk 1]
```diff
@@ -79,12 +79,27 @@ def test_induced_pcgs():
...
     PcGroup = G.polycyclic_group()
     collector = PcGroup.collector
     gens = [G[0], G[1], G[2], G[3]]
     ipcgs = collector.induced_pcgs(gens)
     order = [gen.order() for gen in ipcgs]
-    assert order ==[2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2]
+    assert order == [2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2]
+
+def test_issue_18804():
+    G = DihedralGroup(10)
```
Comment (divyanshu132) [inline sympy/combinatorics/tests/test_pc_groups.py]: There is an existing function to test presentation, this can go to that.

Comment (mohitacecode) [reply sympy/combinatorics/tests/test_pc_groups.py]: can you tell which one I am not able to find it.
or something like this:
```
>>> G = DihedralGroup(10)
>>> a = G.polycyclic_group()
>>> a.is_group
True
```

Comment (divyanshu132) [reply sympy/combinatorics/tests/test_pc_groups.py]: I guess, the first function is to test pc_presentation which has a list containing multiple groups you can append this as well.

Comment (mohitacecode) [reply sympy/combinatorics/tests/test_pc_groups.py]: ohhk got it.

**Comment by mohitacecode** [comment]
I have a query should'nt we test the `induced polycyclic sequence` (ipcgs) by creating the matrix of `exponential vectors` of `xi` in `X(ipcgs)` and checking that whether it in `upper triangular form` or not.(that is the property of ipcgs)

**Comment by divyanshu132** [comment]
Would you like to elaborate with a test example?

**Comment by mohitacecode** [comment]
I am thinking of test like this
```
>>> G = SymmetricGroup(9).sylow_subgroup(3)
>>> PcGroup = G.polycyclic_group()
>>> collector = PcGroup.collector
>>> gens = [G[0], G[1]]
>>> ipcgs = collector.induced_pcgs(gens)
>>> a = []
>>> for i in ipcgs:
...     a.append(collector.exponent_vector(i))
... 
>>> a = Matrix(a)
>>> assert a.is_upper
```

**Comment by divyanshu132** [comment]
This looks ok. I think we can replace the current function `test_induced_pcgs` with something like above.

**Comment by mohitacecode** [comment]
I have changed the test please see if it look ok.

**Comment by mohitacecode** [comment]
the test which is getting timeout running fine on my system.

**Comment by divyanshu132** [comment]
```
>>> SymmetricGroup(20).sylow_subgroup(2).order()
262144
```
I think it's because of this test but the order is not that much huge. May be it's due to the algorithm of induced_pcgs. We should look into GAP and Handbook, if they have described anything about the order of groups till which we can go.

**Comment by mohitacecode** [comment]
It is falling in infinite loop in `_sift` due to the change in condition of while loop.
we are expecting `_sift` to get out of while loop when it found an element whose depth is `n+1` but it is not able to find it in case of `DihedralGroup` It is only able to find element of depth `n` so it is falling in infinite loop there.

In the below example it is only able to find element of depth `3` but we are expecting an element of depth `4` to exit while loop.
for ex:
```
>>> a = DihedralGroup(10)
>>> g = a.polycyclic_group()
>>> collector = g.collector
>>> gens = [a[0],a[1]]
>>> ipcgs = collector.induced_pcgs(gens)
```
But from the algorithm described in handbook we should use `n+1`.

what should be done here. @jksuom @divyanshu132 .

**Comment by divyanshu132** [comment]
Hey sorry for the late reply, Does the loop halt if we are going till depth `n` only.

**Comment by mohitacecode** [comment]
@divyanshu132 Yes It does.

**Comment by divyanshu132** [comment]
Probably, because the indexing in python starts from 0 while in Handbook they consider 1 as the starting index.

File: sympy/combinatorics/tests/test_pc_groups.py [hunk 2]
```diff
@@ -73,18 +73,18 @@ def test_exponent_vector():
...
-
-    G = SymmetricGroup(20).sylow_subgroup(2)
-    PcGroup = G.polycyclic_group()
-    collector = PcGroup.collector
-    gens = [G[0], G[1], G[2], G[3]]
-    ipcgs = collector.induced_pcgs(gens)
-    order = [gen.order() for gen in ipcgs]
-    assert order ==[2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2]
+    G = [SymmetricGroup(9).sylow_subgroup(3), SymmetricGroup(20).sylow_subgroup(2), AlternatingGroup(4),
+    DihedralGroup(4), SymmetricGroup(3), SymmetricGroup(4)]
```
Comment (divyanshu132) [inline sympy/combinatorics/tests/test_pc_groups.py:77]: Can you think of few more examples? So, that we'll be sure that its working as expected.

File: sympy/combinatorics/tests/test_pc_groups.py [hunk 3]
```diff
@@ -73,18 +73,18 @@ def test_exponent_vector():
...
+
+    for g in G:
+        PcGroup = g.polycyclic_group()
+        collector = PcGroup.collector
+        gens = []
+        genr = g.generators
+        for i in range(len(genr)):
+            gens.append(genr[i])
```
Comment (divyanshu132) [inline sympy/combinatorics/tests/test_pc_groups.py:85]: `gens = [gen for gen in g.generators]`

**Comment by mohitacecode** [comment]
@divyanshu132 Sure I have all the suggested changes and also added more test.

**Comment by divyanshu132** [comment]
Looks good!

**Comment by mohitacecode** [comment]
Thanks :)

**Pull request merged**